// react-bootstrap-breadcrumbs/lib/Breadcrumbs.js
// Inspired by
// https://github.com/svenanders/react-breadcrumbs/blob/master/index.js

'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; desc = parent = getter = undefined; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) subClass.__proto__ = superClass; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactRouter = require('react-router');

var _reactRouterLibGetRouteParams = require('react-router/lib/getRouteParams');

var _reactRouterLibGetRouteParams2 = _interopRequireDefault(_reactRouterLibGetRouteParams);

var _reactRouterLibPatternUtils = require('react-router/lib/PatternUtils');

function getBreadcrumbs(routes, routerParams, getTitle) {
  var path = '';
  var params = {};

  var breadcrumbs = routes.map(function (route) {
    params = _extends({}, params, (0, _reactRouterLibGetRouteParams2['default'])(route, routerParams));
    var name = route.name || route.component && (route.component.displayName || route.component.name);

    if (getTitle) {
      var title = getTitle(name, route, params);
      if (title || title === false) {
        name = title;
      }
    }

    if (route.path) {
      if (route.path.indexOf('/') === 0) {
        path = route.path;
      } else {
        path += '/' + route.path;
      }
    }

    return {
      to: path,
      params: params,
      name: name
    };
  }).filter(function (obj) {
    return obj && (obj.name || obj.isActive);
  });

  return breadcrumbs.map(function (obj, idx) {
    return _extends({}, obj, { isActive: idx === breadcrumbs.length - 1 });
  });
}

var Breadcrumbs = (function (_Component) {
  function Breadcrumbs() {
    _classCallCheck(this, Breadcrumbs);

    _get(Object.getPrototypeOf(Breadcrumbs.prototype), 'constructor', this).apply(this, arguments);
  }

  _inherits(Breadcrumbs, _Component);

  _createClass(Breadcrumbs, [{
    key: 'render',
    value: function render() {
      var _props = this.props;
      var getTitle = _props.getTitle;
      var className = _props.className;
      var routes = _props.routes;
      var params = _props.params;

      var breadcrumbs = getBreadcrumbs(routes, params, getTitle);

      return _react2['default'].createElement(
        'ol',
        { className: className ? 'breadcrumb ' + className : 'breadcrumb' },
        breadcrumbs.map(function (breadcrumb, idx) {
          return _react2['default'].createElement(
            'li',
            { className: breadcrumb.isActive && 'active',
              key: 'breadcrumb-' + idx },
            breadcrumb.isActive ? breadcrumb.name || '...' : _react2['default'].createElement(
              _reactRouter.Link,
              { to: (0, _reactRouterLibPatternUtils.formatPattern)(breadcrumb.to, breadcrumb.params) },
              breadcrumb.name
            )
          );
        })
      );
    }
  }], [{
    key: 'propTypes',
    value: {
      getTitle: _react.PropTypes.func,
      className: _react.PropTypes.string,
      routes: _react.PropTypes.array.isRequired,
      params: _react.PropTypes.object
    },
    enumerable: true
  }]);

  return Breadcrumbs;
})(_react.Component);

exports['default'] = Breadcrumbs;
module.exports = exports['default'];
